File: bounce/src/main/java/is/ricko/bounce/config/BounceConfig.java
package is.ricko.bounce.config;

import lombok.Getter;
import lombok.Setter;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Configuration;

import java.util.HashMap;
import java.util.Map;

@Configuration
@Getter
public class BounceConfig {
  public static final String COOKIE_NAME_DEFAULT = "bounce";
  public static final String DOMAIN_DEFAULT = "ricko.is";
  public static final String GITHUB = "github";
  public static final String GOOGLE = "google";

  @Setter(onMethod = @__(@Value("${is.ricko.bounce.cookie.comment:Correlate multiple ricko.is links - never shared}")))
  private String cookieComment = "Correlate multiple ricko.is links - never shared";

  private int cookieMaxAge = 2147483647;

  @Setter(onMethod = @__(@Value("${is.ricko.bounce.cookie.name:" + COOKIE_NAME_DEFAULT + "}")))
  private String cookieName = COOKIE_NAME_DEFAULT;

  @Setter(onMethod = @__(@Value("${is.ricko.bounce.domain:" + DOMAIN_DEFAULT + "}")))
  private String domain = DOMAIN_DEFAULT;

  @Setter(onMethod = @__(@Value("${is.ricko.bounce.jdbc.password}")))
  private String jdbcPassword = null;

  @Setter(onMethod = @__(@Value("${is.ricko.bounce.jdbc.url}")))
  private String jdbcUrl = null;

  @Setter(onMethod = @__(@Value("${is.ricko.bounce.jdbc.username}")))
  private String jdbcUsername = null;

  private boolean logHeaders = false;

  private final Map<String, String> oauth2clientId = new HashMap<>();

  private final Map<String, String> oauth2clientSecret = new HashMap<>();

  private String oauth2redirectUri = "{baseUrl}";

  public String getRedirectUri() {
    return oauth2redirectUri;
  }

  @Value("${is.ricko.bounce.oauth.github.clientId:}")
  public void setGitHubClientId(final String clientId) {
    setIfNotNull(oauth2clientId, GITHUB, clientId);
  }

  @Value("${is.ricko.bounce.oauth.github.clientSecret:}")
  public void setGitHubClientSecret(final String secret) {
    setIfNotNull(oauth2clientId, GITHUB, secret);
  }

  @Value("${is.ricko.bounce.oauth.google.clientId:}")
  public void setGoogleClientId(final String clientId) {
    setIfNotNull(oauth2clientId, GOOGLE, clientId);
  }

  @Value("${is.ricko.bounce.oauth.google.clientSecret:}")
  public void setGoogleClientSecret(final String secret) {
    setIfNotNull(oauth2clientSecret, GOOGLE, secret);
  }

  public String clientIdFor(final String client) {
    return oauth2clientId.get(client);
  }

  public String secretFor(final String client) {
    return oauth2clientSecret.get(client);
  }

  private void setIfNotNull(final Map<String, String> map, final String key, final String value) {
    if (value != null && !value.isEmpty()) {
      map.put(key, value);
    }
  }

  @Value("${is.ricko.bounce.log.headers:}")
  public void setLogHeaders(final String logHeaders) {
    if (logHeaders != null && !logHeaders.isEmpty()) {
      this.logHeaders = Boolean.valueOf(logHeaders);
    }
  }

  @Value("${is.ricko.bounce.cookie.maxAge:2147483647}")
  public void setMaxAgeFromString(final String maybeMaxAge) {
    if (maybeMaxAge != null) {
      cookieMaxAge = Integer.valueOf(maybeMaxAge, 10);
    }
  }

  @Value("${is.ricko.bounce.oauth.redirect:}")
  public void setOAuthRedirect(final String redirectUri) {
    if (redirectUri != null && !redirectUri.isEmpty()) {
      oauth2redirectUri = redirectUri;
    }
  }
}

File: bounce/src/main/kotlin/is/ricko/bounce/BounceConfig.kt
package `is`.ricko.bounce

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Configuration

@Configuration
class BounceConfig {
    var cookieComment: String = "Correlate multiple ricko.is links - never shared"
        @Value("\${is.ricko.bounce.cookie.comment:Correlate multiple ricko.is links - never shared}") set

    var cookieName: String = "bounce"
        @Value("\${is.ricko.bounce.cookie.name:bounce}") set

    var domain: String = "ricko.is"
        @Value("\${is.ricko.bounce.domain:ricko.is}") set

    var cookieMaxAge: Int = 2147483647

    var jdbcUrl: String? = null
        @Value("\${is.ricko.bounce.jdbc.url}") set

    var jdbcUsername: String? = null
        @Value("\${is.ricko.bounce.jdbc.username}") set

    var jdbcPassword: String? = null
        @Value("\${is.ricko.bounce.jdbc.password}") set

    val oauth2clientId: MutableMap<String, String> = HashMap()

    val oauth2clientSecret: MutableMap<String, String> = HashMap()

    var oauth2redirectUri: String = "{baseUrl}"

    @Value("\${is.ricko.bounce.cookie.maxAge:2147483647}")
    fun setMaxAgeFromString(maybeMaxAge: String?) {
        if (maybeMaxAge != null) {
            cookieMaxAge = maybeMaxAge.toInt()
        }
    }

    @Value("\${is.ricko.bounce.oauth.google.clientId:}")
    fun setGoogleClientId(clientId: String?) {
        setIfNotNull(oauth2clientId, GOOGLE, clientId)
    }

    @Value("\${is.ricko.bounce.oauth.github.clientId:}")
    fun setGitHubClientId(clientId: String?) {
        setIfNotNull(oauth2clientId, GITHUB, clientId)
    }

    @Value("\${is.ricko.bounce.oauth.google.clientSecret:}")
    fun setGoogleClientSecret(secret: String?) {
        setIfNotNull(oauth2clientSecret, GOOGLE, secret)
    }

    @Value("\${is.ricko.bounce.oauth.github.clientSecret:}")
    fun setGitHubClientSecret(secret: String?) {
        setIfNotNull(oauth2clientId, GITHUB, secret)
    }

    @Value("\${is.ricko.bounce.oauth.redirect:}")
    fun setOAuthRedirect(redirectUri: String?) {
        if (redirectUri != null && !redirectUri.isBlank()) {
            oauth2redirectUri = redirectUri
        }
    }

    private fun setIfNotNull(map: MutableMap<String, String>, client: String, value: String?) {
        if (value != null && !value.isBlank()) {
            map[client] = value
        }
    }
}

const val GOOGLE = "google"

const val GITHUB = "github"